activerecord/lib/active_record/transactions.rb

```diff
@@ -356,6 +356,7 @@ def add_to_transaction
       if has_transactional_callbacks?
         self.class.connection.add_transaction_record(self)
       else
+        sync_with_transaction_state
```

Comment:
It won't clean up state unless two conditions are met:
- a transaction corresponding to transaction_state was finalized (https://github.com/rails/rails/blob/master/activerecord/lib/active_record/transactions.rb#L478)
- a transaction level has reached 0 (https://github.com/rails/rails/blob/master/activerecord/lib/active_record/transactions.rb#L414, https://github.com/rails/rails/blob/master/activerecord/lib/active_record/transactions.rb#L402)

So it looks like a safe method to call. It will do appropriate job when necessary.